Gems/Atom/RHI/DX12/Code/Source/RHI/DescriptorContext.cpp

```diff
@@ -432,14 +514,78 @@ namespace AZ
         {
             AZ_Assert(type < D3D12_DESCRIPTOR_HEAP_TYPE_NUM_TYPES, "Trying to get pool with invalid type: [%d]", type);
             AZ_Assert(flag < NumHeapFlags, "Trying to get pool with invalid flag: [%d]", flag);
-            return m_pools[type][flag];
+
+            if (m_allowDescriptorHeapCompaction && IsShaderVisibleCbvSrvUavHeap(type, flag))
+            {
+                return m_shaderVisibleCbvSrvUavPools[m_currentHeapIndex];
+            }
+            else
+            {
+                return m_pools[type][flag];
+            }
         }
 
         const DescriptorPool& DescriptorContext::GetPool(uint32_t type, uint32_t flag) const
         {
             AZ_Assert(type < D3D12_DESCRIPTOR_HEAP_TYPE_NUM_TYPES, "Trying to get pool with invalid type: [%d]", type);
             AZ_Assert(flag < NumHeapFlags, "Trying to get pool with invalid flag: [%d]", flag);
-            return m_pools[type][flag];
+            if (m_allowDescriptorHeapCompaction && IsShaderVisibleCbvSrvUavHeap(type, flag))
+            {
+                return m_shaderVisibleCbvSrvUavPools[m_currentHeapIndex];
+            }
+            else
+            {
+                return m_pools[type][flag];
+            }
+        }
+
+        void DescriptorContext::CompactDescriptorHeap()
+        {
+            //Check if heap compaction is enabled by the used. Since there is an overhead associated with heap compaction it is not enabled by default
+            if(!m_allowDescriptorHeapCompaction)
+            {
+                AZ_Assert(
+                    false,
+                    "Descriptor heap Compaction not allowed. Please consider increasing number of handles allowed for the second value"
+                    "of DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV or enabling AllowDescriptorHeapCompaction within platformlimits.azasset file for dx12.");
+                return;
+            }
+
+            //We need to ping-pong between two heaps as we cannot compact the active heap without updating it and that is not allowed as
+            //we need to keep that gpu memory untouched until GPU is finished consuming which can take up to 3 frames. 
+            m_compactionInProgress = true;
```

Comment:
I think there is a thread safety issue here. Suppose we invoke `DescriptorContext::CreateDescriptorTable` and start running _after_ checking the `m_compactionInProgress` flag and then invoke this code. This code never grabs that mutex guarding `m_srgAllocations` and the code will likely crash or corrupt memory.

Reply:
There shouldnt be a threading issue as this function is called once all the SRG compilations across various threads are finished. I can ask a mutex just because someone in future can introduce a bug maybe?